File: Bronze1-1/06-24-am/unit09/adding-parentheses/parentheses.cpp
/*
 * Adding parentheses
 */
#include <bits/stdc++.h>
using namespace std;

vector<int> solve(string str) {
	vector<int> ret;

	int len = str.length();
    for(int j=0; j<len; j++) {
        char c = str[j];
        if(c == '*' || c == '-' || c == '+') {
            vector<int> left = solve(str.substr(0, j));
            vector<int> right = solve(str.substr(j+1, len-j-1));
            for(int x : left) {
                for(int y : right) {
                	if(c == '+' && find(ret.begin(), ret.end(), x+y)==ret.end())
                    	ret.push_back(x+y);
                    else if(c == '-' && find(ret.begin(), ret.end(), x-y)==ret.end())
                    	ret.push_back(x-y);
                    else if(c == '*' && find(ret.begin(), ret.end(), x*y)==ret.end())
                    	ret.push_back(x*y);
                }
            }
        }
    }
    
    if(ret.empty()) ret.push_back(stoi(str));
    
    return ret;
}

int main() {
	string str; cin >> str;
	
	vector<int> ans = solve(str);
	sort(ans.begin(), ans.end());
	for(int v : ans)
		cout << v << " ";
	cout << '\n';
}

File: Bronze1-1/06-24-am/unit09/adding-parentheses/parentheses.java
/*
 * Adding parentheses
 */
import java.util.*;
public class parentheses {
	public static void main(String[] args) {
		Scanner in = new Scanner(System.in);
		String str = in.nextLine();
		
		ArrayList<Integer> ans = solve(str);
		Collections.sort(ans);
		for(int v : ans)
			System.out.print(v + " ");
		System.out.println();
		in.close();
	}
	
	static ArrayList<Integer> solve(String str) {
		ArrayList<Integer> ret = new ArrayList<>();

		int len = str.length();
        for(int j=0; j<len; j++) {
            char c = str.charAt(j);
            if(c == '*' || c == '-' || c == '+') {
                ArrayList<Integer> left = solve(str.substring(0, j));
                ArrayList<Integer> right = solve(str.substring(j+1, len));
                for(Integer x : left) {
                    for(Integer y : right) {
                    	if(c == '+' && !ret.contains(x+y)) 
                        	ret.add(x + y);
                        else if(c == '-' && !ret.contains(x-y)) 
                        	ret.add(x - y);
                        else if(c == '*' && !ret.contains(x*y)) 
                        	ret.add(x * y);
                    }
                }
            }
        }
        
        if(ret.isEmpty()) ret.add(Integer.parseInt(str));
        
        return ret;
	}
}